rust/src/delta.rs

```diff
@@ -613,6 +673,127 @@ impl std::fmt::Debug for DeltaTable {
 
+#[derive(thiserror::Error, Debug)]
+pub enum DeltaTransactionError {
+    #[error("Version already exists")]
+    VersionExists,
+
+    #[error("Maximum number of retries exceeded for committing the transaction")]
+    CommitRetriesExceeded,
+
+    #[error("RecordBatch is missing partition column in Delta schema.")]
+    MissingPartitionColumn,
+
+    #[error("Commit failed: {source}")]
+    Storage { source: StorageError },
+
+    #[error("DeltaTable interaction failed: {source}")]
+    DeltaTable { source: DeltaTableError },
+}
+
+impl From<StorageError> for DeltaTransactionError {
+    fn from(error: StorageError) -> Self {
+        match error {
+            StorageError::AlreadyExists => DeltaTransactionError::VersionExists,
+            _ => DeltaTransactionError::Storage { source: error },
+        }
+    }
+}
+
+impl From<DeltaTableError> for DeltaTransactionError {
+    fn from(error: DeltaTableError) -> Self {
+        DeltaTransactionError::DeltaTable { source: error }
+    }
+}
+
+pub struct DeltaTransaction<'a> {
+    delta_table: &'a mut DeltaTable,
+}
+
+impl<'a> DeltaTransaction<'a> {
+    pub fn new(delta_table: &'a mut DeltaTable) -> Self {
+        DeltaTransaction { delta_table,}
+    }
+
+    pub async fn commit_all(&mut self, actions: &[Action], _operation: Option<DeltaOperation>) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        // TODO: stubbing `operation` parameter (which will be necessary for writing the CommitInfo action), but leaving it unused for now.
+        // `CommitInfo` is a fairly dynamic data structure so we should work out the data structure approach separately.
+
+        // TODO: calculate isolation level to use when checking for conflicts.
+        // Leaving conflict checking unimplemented for now to get the "single writer" implementation off the ground.
+        // Leaving some commmented code in place as a guidepost for the future.
+
+        // let no_data_changed = actions.iter().all(|a| match a {
+        //     Action::add(x) => !x.dataChange,
+        //     Action::remove(x) => !x.dataChange,
+        //     _ => false,
+        // });
+        // let isolation_level = if no_data_changed {
+        //     IsolationLevel::SnapshotIsolation
+        // } else { 
+        //     IsolationLevel::Serializable
+        // };
+
+        // TODO: create a CommitInfo action and prepend it to actions.
+
+        // try to commit in a loop in case other writers write the next version first
+        let version = self.try_commit_loop(actions).await?;
+
+        self.delta_table.update().await?;
+
+        Ok(version)
+    }
+
+    async fn try_commit_loop(&mut self, actions: &[Action]) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        let max_retries = max_retry_commit_attempts();
+        let mut attempt_number: u32 = 0;
+
+        loop {
+            if attempt_number > max_retries {
+                return Err(DeltaTransactionError::CommitRetriesExceeded);
+            }
```

Comment:
nit pick, we can move this to the bottom of the loop after incremented attempt_number. the first check entering the loop is an no op because it will always be true.

Reply:
fixed

Reply:
newly committed changes cut across a few different comments. might be best to start from https://github.com/delta-io/delta-rs/pull/84/files#diff-3b5c80af9e5527fbeaa6dad151a835f0733a26e61d0ba0471c58f243a29dfa99R796 for re-review.